<div style="border:solid black 2px; padding: 20px"><h1 style="color:black; margin-bottom:20px">Reviewer's comment v1</h1>
Halo Mufti !

Perkenalkan saya Dipta, disini saya akan mereview tugas Mufti ya.

Saya akan memberikan beberapa komen seperti dibawah ini, tolong jangan dipindah, dirubah, maupun dihapus ya.

Komen yang saya berikan akan muncul dalam warna hijau, kuning, atau pun merah seperti ini:

<div class="alert alert-success">
<b>Reviewer's comment</b> <a class="tocSkip"></a>

Bagus, semua berjalan lancar.

</div>

<div class="alert alert-warning">
<b>Reviewer's comment</b> <a class="tocSkip"></a>

Terdapat beberapa catatan.

</div>

<div class="alert alert-danger">
<b>Reviewer's comment</b> <a class="tocSkip"></a>

Perlu beberapa perbaikan

</div>

Dan tentu saja Mufti dapat menjawab saya dengan menggunakan ini

<div class="alert alert-block alert-info">
<b>Student answer.</b> <a class="tocSkip"></a>
    
Baik kak
</div>

# Menganalisis Risiko Gagal Bayar Peminjam

Tugas Anda adalah menyiapkan laporan untuk divisi kredit suatu bank. Anda akan mencari tahu pengaruh status perkawinan seorang nasabah dan jumlah anak yang dimilikinya terhadap probabilitas gagal bayar dalam pelunasan pinjaman. Pihak bank sudah memiliki beberapa data mengenai kelayakan kredit nasabah.

Laporan Anda akan dipertimbangkan pada saat membuat **penilaian kredit** untuk calon nasabah. **Penilaian kredit** digunakan untuk mengevaluasi kemampuan calon peminjam untuk melunasi pinjaman mereka.

<div class="alert alert-success">
<b>Reviewer's comment</b> <a class="tocSkip"></a>

Bagus, sudah diberikan terkait deskripsi dari projectnya.

</div>

## Buka *file* data dan baca informasi umumnya.

In [186]:
# Muat semua *library*
import pandas as pd

# Muat datanya
df = pd.read_csv('/datasets/credit_scoring_eng.csv')

<div class="alert alert-success">
<b>Reviewer's comment</b> <a class="tocSkip"></a>

Bagus, semua berjalan lancar.

</div>

## Soal 1. Eksplorasi data

**Deskripsi Data**
- `children` - jumlah anak dalam keluarga
- `days_employed` - pengalaman kerja nasabah dalam hari
- `dob_years` - usia nasabah dalam tahun
- `education` - tingkat pendidikan nasabah
- `education_id` - pengidentifikasi untuk tingkat pendidikan nasabah
- `family_status` - pengidentifikasi untuk status perkawinan nasabah
- `family_status_id` - tanda pengenal status perkawinan
- `gender` - jenis kelamin nasabah
- `income_type` - jenis pekerjaan
- `debt` - apakah nasabah memiliki hutang pembayaran pinjaman
- `total_income` - pendapatan bulanan
- `purpose` - tujuan mendapatkan pinjaman

In [187]:
# Mari kita lihat berapa banyak baris dan kolom yang dimiliki oleh dataset kita
df.shape

#-Ada 21525 Baris, dan 12 Kolom

(21525, 12)

In [188]:
# Mari tampilkan N baris pertama
df.head(20)


children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42    bachelor's degree             0   
1          1   -4024.803754         36  secondary education             1   
2          0   -5623.422610         33  Secondary Education             1   
3          3   -4124.747207         32  secondary education             1   
4          0  340266.072047         53  secondary education             1   
5          0    -926.185831         27    bachelor's degree             0   
6          0   -2879.202052         43    bachelor's degree             0   
7          0    -152.779569         50  SECONDARY EDUCATION             1   
8          2   -6929.865299         35    BACHELOR'S DEGREE             0   
9          0   -2188.756445         41  secondary education             1   
10         2   -4171.483647         36    bachelor's degree             0   
11         0    -792.701887         40  secondary education             1   
12         0            NaN         65  secondary education             1   
13         0   -1846.641941         54         some college             2   
14         0   -1844.956182         56    bachelor's degree             0   
15         1    -972.364419         26  secondary education             1   
16         0   -1719.934226         35  secondary education             1   
17         0   -2369.999720         33    bachelor's degree             0   
18         0  400281.136913         53  secondary education             1   
19         0  -10038.818549         48  SECONDARY EDUCATION             1   

        family_status  family_status_id gender income_type  debt  \
0             married                 0      F    employee     0   
1             married                 0      F    employee     0   
2             married                 0      M    employee     0   
3             married                 0      M    employee     0   
4   civil partnership                 1      F     retiree     0   
5   civil partnership                 1      M    business     0   
6             married                 0      F    business     0   
7             married                 0      M    employee     0   
8   civil partnership                 1      F    employee     0   
9             married                 0      M    employee     0   
10            married                 0      M    business     0   
11            married                 0      F    employee     0   
12  civil partnership                 1      M     retiree     0   
13            married                 0      F    employee     0   
14  civil partnership                 1      F    business     1   
15            married                 0      F    employee     0   
16            married                 0      F    employee     0   
17  civil partnership                 1      M    employee     0   
18    widow / widower                 2      F     retiree     0   
19           divorced                 3      F    employee     0   

    total_income                              purpose  
0      40620.102                purchase of the house  
1      17932.802                         car purchase  
2      23341.752                purchase of the house  
3      42820.568              supplementary education  
4      25378.572                    to have a wedding  
5      40922.170                purchase of the house  
6      38484.156                 housing transactions  
7      21731.829                            education  
8      15337.093                     having a wedding  
9      23108.150  purchase of the house for my family  
10     18230.959                      buy real estate  
11     12331.077           buy commercial real estate  
12           NaN                    to have a wedding  
13     20873.317                         car purchase  
14     26420.466          buy residential real estate  
15     18691.345         construction of own property  
16   

<div class="alert alert-block alert-info">Berdasarkan tabel sampel diatas yang menampilkan 20 baris, berikut beberapa masalah yang mungkin memerlukan penyelidikan lebih lanjut.
- Kolom 'days_employed' masih banyak yang bernilai negatif. Ada 1 baris yang bernilai 'NaN'. Dan semua baris masih bernilai float.
- Kolom 'education' mempunya nilai duplikat, contohnya mengandung arti yang sama namum beda penulisan (Huruf Besar &    Huruf Kecil).
- Kolom 'total_income' ada 1 baris yang bernilai 'NaN' dan sejajar dengan 'NaN' di 'days_employed'. Semua baris masih bernilai float.<a class="tocSkip"></a></div>

In [189]:
# Dapatkan informasi data
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


<div class="alert alert-block alert-info">Berdasarkan metode info() diatas terdapat nilai yang hilang di kolom 'days_employed' dan 'total_income'. Jumlah nilai yang hilang di kedua kolom tersebut sama.</br></br><a class="tocSkip">

In [190]:
# Mari kita lihat tabel yang difilter dengan nilai yang hilang di kolom pertama yang mengandung data yang hilang
df[df['days_employed'].isna()]


children  days_employed  dob_years            education  education_id  \
12            0            NaN         65  secondary education             1   
26            0            NaN         41  secondary education             1   
29            0            NaN         63  secondary education             1   
41            0            NaN         50  secondary education             1   
55            0            NaN         54  secondary education             1   
...         ...            ...        ...                  ...           ...   
21489         2            NaN         47  Secondary Education             1   
21495         1            NaN         50  secondary education             1   
21497         0            NaN         48    BACHELOR'S DEGREE             0   
21502         1            NaN         42  secondary education             1   
21510         2            NaN         28  secondary education             1   

           family_status  family_status_id gender    income_type  debt  \
12     civil partnership                 1      M        retiree     0   
26               married                 0      M  civil servant     0   
29             unmarried                 4      F        retiree     0   
41               married                 0      F  civil servant     0   
55     civil partnership                 1      F        retiree     1   
...                  ...               ...    ...            ...   ...   
21489            married                 0      M       business     0   
21495  civil partnership                 1      F       employee     0   
21497            married                 0      F       business     0   
21502            married                 0      F       employee     0   
21510            married                 0      F       employee     0   

       total_income                   purpose  
12              NaN         to have a wedding  
26              NaN                 education  
29              NaN    building a real estate  
41              NaN  second-hand car purchase  
55              NaN         to have a wedding  
...             ...                       ...  
21489           NaN         purchase of a car  
21495           NaN          wedding ceremony  
21497           NaN       building a property  
21502           NaN    building a real estate  
21510           NaN              car purchase  

[2174 rows x 12 columns]

<div class="alert alert-block alert-info">Berdasarkan filter dengan kolom 'days_employed' dengan nilai yang hilang, tampak nilai hilang tersebut simetris pada kolom 'total_income'. Namun ini masih membutuhkan penyelidikan lebih lanjut.<a class="tocSkip"></a></div>

In [191]:
# Mari kita terapkan beberapa kondisi untuk memfilter data dan melihat jumlah baris dalam tabel yang telah difilter.
df.isnull().sum()


children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

**Kesimpulan sementara**

<div class="alert alert-block alert-info">Jumlah baris tabel yang telah difilter sesuai dengan jumlah nilai yang hilaing. Berarti hilangnya nilai tersebut terdapat pola atau tidak random.<a class="tocSkip"></a></div>

<div class="alert alert-block alert-info">Jumlah baris yang hilang nilainya yaitu 21525 - 19351 = 2174 yaitu sekitar 10%.<a class="tocSkip"></a></div>

<div class="alert alert-block alert-info">
- Kolom 'days_employed' dan 'total_income' adalah variabel kuantitafif.
- Nilai yang hilang untuk variabel tersebut bisa disi dengan nilai representatif.
- Nilai representatif tersebut bisa menggunakan median atau mean dari masin-masing kolom tersebut.<a class="tocSkip"></a></div>


<div class="alert alert-success">
<b>Reviewer's comment</b> <a class="tocSkip"></a>

Bagus, semua berjalan lancar.

</div>

In [192]:
# Mari kita periksa nasabah yang tidak memiliki data tentang karakteristik yang teridentifikasi dan kolom dengan nilai yang hilang
#df_notnull = df[df['days_employed'].notnull()]
#df_notnull

df_isnull = df[df['days_employed'].isnull()]
df_isnull


children  days_employed  dob_years            education  education_id  \
12            0            NaN         65  secondary education             1   
26            0            NaN         41  secondary education             1   
29            0            NaN         63  secondary education             1   
41            0            NaN         50  secondary education             1   
55            0            NaN         54  secondary education             1   
...         ...            ...        ...                  ...           ...   
21489         2            NaN         47  Secondary Education             1   
21495         1            NaN         50  secondary education             1   
21497         0            NaN         48    BACHELOR'S DEGREE             0   
21502         1            NaN         42  secondary education             1   
21510         2            NaN         28  secondary education             1   

           family_status  family_status_id gender    income_type  debt  \
12     civil partnership                 1      M        retiree     0   
26               married                 0      M  civil servant     0   
29             unmarried                 4      F        retiree     0   
41               married                 0      F  civil servant     0   
55     civil partnership                 1      F        retiree     1   
...                  ...               ...    ...            ...   ...   
21489            married                 0      M       business     0   
21495  civil partnership                 1      F       employee     0   
21497            married                 0      F       business     0   
21502            married                 0      F       employee     0   
21510            married                 0      F       employee     0   

       total_income                   purpose  
12              NaN         to have a wedding  
26              NaN                 education  
29              NaN    building a real estate  
41              NaN  second-hand car purchase  
55              NaN         to have a wedding  
...             ...                       ...  
21489           NaN         purchase of a car  
21495           NaN          wedding ceremony  
21497           NaN       building a property  
21502           NaN    building a real estate  
21510           NaN              car purchase  

[2174 rows x 12 columns]

<div class="alert alert-danger">
<b>Reviewer's comment</b> <a class="tocSkip"></a>

ini harusnya difilter yang is null ya mas, bukan yg not null. setelah itu baru dibandingkan dengan seluruh dataset

</div>

<div class="alert alert-block alert-info">
<b>Student answer.</b> <a class="tocSkip"></a>
    
Terima Kasih Untuk Mas Dipta untuk Reviewnya. Sudah diperbaiki, dari tabel dengan metode isnull() tersebut dapat disimpulkan sementera untuk nilai yang hilang polanya tidak acak, karena bila kolom
'days_employed' bernilai NaN maka kolom 'total_incomenya' juga NaN.
</div>

In [193]:
# Periksalah distribusinya
#df_notnull.describe()
df_isnull.describe()

children  days_employed    dob_years  education_id  \
count  2174.000000            0.0  2174.000000   2174.000000   
mean      0.552438            NaN    43.632015      0.800828   
std       1.469356            NaN    12.531481      0.530157   
min      -1.000000            NaN     0.000000      0.000000   
25%       0.000000            NaN    34.000000      0.250000   
50%       0.000000            NaN    43.000000      1.000000   
75%       1.000000            NaN    54.000000      1.000000   
max      20.000000            NaN    73.000000      3.000000   

       family_status_id         debt  total_income  
count       2174.000000  2174.000000           0.0  
mean           0.975161     0.078197           NaN  
std            1.418220     0.268543           NaN  
min            0.000000     0.000000           NaN  
25%            0.000000     0.000000           NaN  
50%            0.000000     0.000000           NaN  
75%            1.000000     0.000000           NaN  
max            4.000000     1.000000           NaN

**Kemungkinan penyebab hilangnya nilai dalam data**

In [194]:
# Memeriksa distribusi di seluruh *dataset*
df.describe()

children  days_employed     dob_years  education_id  \
count  21525.000000   19351.000000  21525.000000  21525.000000   
mean       0.538908   63046.497661     43.293380      0.817236   
std        1.381587  140827.311974     12.574584      0.548138   
min       -1.000000  -18388.949901      0.000000      0.000000   
25%        0.000000   -2747.423625     33.000000      1.000000   
50%        0.000000   -1203.369529     42.000000      1.000000   
75%        1.000000    -291.095954     53.000000      1.000000   
max       20.000000  401755.400475     75.000000      4.000000   

       family_status_id          debt   total_income  
count      21525.000000  21525.000000   19351.000000  
mean           0.972544      0.080883   26787.568355  
std            1.420324      0.272661   16475.450632  
min            0.000000      0.000000    3306.762000  
25%            0.000000      0.000000   16488.504500  
50%            0.000000      0.000000   23202.870000  
75%            1.000000      0.000000   32549.611000  
max            4.000000      1.000000  362496.645000

**Kesimpulan sementara**
<div class="alert alert-block alert-info">Berdasarkan tabel df_notnull (tabel dengan filter tidak dengan nilai null) dan df(tabel asli) distribusinya mirip. Berarti jika kita menghapus/mengdrop baris yang bernilai null seharusnya tidak terlalu berpengaruh.<a class="tocSkip"></a></div>

In [195]:
# Periksa penyebab dan pola lain yang dapat mengakibatkan nilai yang hilang
print("Menghitung nilai hilang 'days_employed' berdasarkan 'gender'")
print(df[df['days_employed'].isnull()]['gender'].value_counts())

print()
print("Menghitung nilai hilang 'total_income' berdasarkan 'gender'")
print(df[df['total_income'].isnull()]['gender'].value_counts())


Menghitung nilai hilang 'days_employed' berdasarkan 'gender'
F    1484
M     690
Name: gender, dtype: int64

Menghitung nilai hilang 'total_income' berdasarkan 'gender'
F    1484
M     690
Name: gender, dtype: int64


<div class="alert alert-success">
<b>Reviewer's comment</b> <a class="tocSkip"></a>

Bagus, semua berjalan lancar.

</div>

**Kesimpulan sementara**
<div class="alert alert-block alert-info">Dari hitungan diatas yang menghasilkan jumlah nilai hilang berdasarkan gender, dapat disimpulkan bahwa hilangnya nilai tersebut tidak terjadi secara acak.<a class="tocSkip"></a></div>

In [196]:
# Periksa pola lainnya - jelaskan pola tersebut

<div class="alert alert-block alert-info">Semua kesimpulan sementara diatas baru berdasarkan hal berikut :
    
- Tampilan 20 Baris pertama pada Dataframe.
- Penggunaan metode info() dan shape untuk informasi umum Dataframe.
- Penggunaan isna(), isnull() untuk pemfilteran tabel dengan nilai yang hilang.
- Penggunaan describe() untuk statitiska deskriptif Dataframe.

Kita perlu memperbaiki nilai hilang tersebut dan memperbaiki Dataframe lebih lanjut dengan melakukan Transformasi Data dan Pengkategorian Data.<a class="tocSkip"></a></div>

## Transformasi data

In [197]:
# Mari kita lihat semua nilai di kolom pendidikan untuk memeriksa ejaan apa yang perlu diperbaiki
df['education'].sort_values().unique()


array(["BACHELOR'S DEGREE", "Bachelor's Degree", 'GRADUATE DEGREE',
       'Graduate Degree', 'PRIMARY EDUCATION', 'Primary Education',
       'SECONDARY EDUCATION', 'SOME COLLEGE', 'Secondary Education',
       'Some College', "bachelor's degree", 'graduate degree',
       'primary education', 'secondary education', 'some college'],
      dtype=object)

In [198]:
# Perbaiki pencatatan jika diperlukan
df['education'] = df['education'].str.lower()


In [199]:
# Periksa semua nilai di kolom untuk memastikan bahwa kita telah memperbaikinya dengan tepat
df['education'].value_counts()


secondary education    15233
bachelor's degree       5260
some college             744
primary education        282
graduate degree            6
Name: education, dtype: int64

<div class="alert alert-success">
<b>Reviewer's comment</b> <a class="tocSkip"></a>

Bagus, semua berjalan lancar.

</div>

In [200]:
# Mari kita lihat distribusi nilai pada kolom `children`
df['children'].sort_values().unique()

array([-1,  0,  1,  2,  3,  4,  5, 20])

<div class="alert alert-block alert-info">Berikut beberapa penjelasan untuk hal yang aneh untuk nilai di kolom 'children'.
- Terdapat nilai -1, kemungkinan ini adalah typo, kita bisa mengganti nilai tersebut menjadi 1 menggunakan metode .replace()
- Mungkin kita melihat hal yang aneh pada nilai 20, apakah benar nasabah tersebut mempunyai 20 anak ? Namun berdasarkan studi seorang wanita bisa mempunya maksimal 15-30 anak untuk 15 kehamilan. 
    <a class="tocSkip"></a></div>

In [201]:
# [perbaiki data berdasarkan keputusan Anda]
df['children'] = df['children'].replace(-1, 1)

In [202]:
# Periksa kembali kolom `children` untuk memastikan bahwa semuanya telah diperbaiki
df['children'].value_counts()

0     14149
1      4865
2      2055
3       330
20       76
4        41
5         9
Name: children, dtype: int64

<div class="alert alert-success">
<b>Reviewer's comment</b> <a class="tocSkip"></a>

jadi yang nilai 20 tidak diganti, oke tidak apa2 karena sudah diberikan alasan bahwa 1 orang wanita bisa punya 15-30 anak.

</div>

In [203]:
# Temukan data yang bermasalah di `days_employed`, jika memang terdapat masalah, dan hitung persentasenya
df['days_employed'].sort_values().unique()


array([-18388.94990057, -17615.56326563, -16593.47281726, ...,
       401715.81174889, 401755.40047533,             nan])

In [204]:
days_employed_null = df['days_employed'].isnull().sum()/df.shape[0]
print('Jumlah percentage untuk nilai yang hilang pada kolom days_employed adalah {:.2%}'.format(days_employed_null))

Jumlah percentage untuk nilai yang hilang pada kolom days_employed adalah 10.10%


In [205]:
df['days_employed'] = abs(df['days_employed'])

<div class="alert alert-block alert-info">Pada saat kesimpulan sementara di bagian Explorasi Data, sudah dijelaskan beberapa masalah pada kolom 'days_employed' :
- Beberapa baris bernilai negatif, hal ini perlu diperbaiki dengan menggunakan fungsi abs().
- Untuk baris yang bernilai 'NaN' akan diperbaiki pada bagian 'Bekerja dengan nilai yang hilang'.
- Semua baris masih bertipe float, hal ini perlu diperbaiki menggunakan astype() setelah memperbaiki Nilai 'NaN'.<a class="tocSkip"></a></div>

<div class="alert alert-success">
<b>Reviewer's comment</b> <a class="tocSkip"></a>

Bagus, semua berjalan lancar.

</div>

In [206]:
# Periksa hasilnya - pastikan bahwa masalahnya telah diperbaiki
df['days_employed'].head(10)

0      8437.673028
1      4024.803754
2      5623.422610
3      4124.747207
4    340266.072047
5       926.185831
6      2879.202052
7       152.779569
8      6929.865299
9      2188.756445
Name: days_employed, dtype: float64

In [207]:
df['dob_years'].max()

75

In [208]:
340266.072047 / 365

932.2358138273972

In [209]:
(df['days_employed'] / 365).describe()
#df_year_employed.describe()

count    19351.000000
mean       183.328024
std        380.906522
min          0.066141
25%          2.539751
50%          6.011563
75%         15.172281
max       1100.699727
Name: days_employed, dtype: float64

In [210]:
df['days_employed'].describe()
#df.head(10)

count     19351.000000
mean      66914.728907
std      139030.880527
min          24.141633
25%         927.009265
50%        2194.220567
75%        5537.882441
max      401755.400475
Name: days_employed, dtype: float64

In [211]:
df.sort_values(by='days_employed' , ascending=False)


children  days_employed  dob_years            education  education_id  \
6954          0  401755.400475         56  secondary education             1   
10006         0  401715.811749         69    bachelor's degree             0   
7664          1  401675.093434         61  secondary education             1   
2156          0  401674.466633         60  secondary education             1   
7794          0  401663.850046         61  secondary education             1   
...         ...            ...        ...                  ...           ...   
21489         2            NaN         47  secondary education             1   
21495         1            NaN         50  secondary education             1   
21497         0            NaN         48    bachelor's degree             0   
21502         1            NaN         42  secondary education             1   
21510         2            NaN         28  secondary education             1   

           family_status  family_status_id gender income_type  debt  \
6954     widow / widower                 2      F     retiree     0   
10006          unmarried                 4      F     retiree     0   
7664             married                 0      F     retiree     0   
2156             married                 0      M     retiree     0   
7794   civil partnership                 1      F     retiree     0   
...                  ...               ...    ...         ...   ...   
21489            married                 0      M    business     0   
21495  civil partnership                 1      F    employee     0   
21497            married                 0      F    business     0   
21502            married                 0      F    employee     0   
21510            married                 0      F    employee     0   

       total_income                 purpose  
6954      28204.551      housing renovation  
10006      9182.441    getting an education  
7664      20194.323    housing transactions  
2156      52063.316                    cars  
7794       7725.831        wedding ceremony  
...             ...                     ...  
21489           NaN       purchase of a car  
21495           NaN        wedding ceremony  
21497           NaN     building a property  
21502           NaN  building a real estate  
21510           NaN            car purchase  

[21525 rows x 12 columns]

In [212]:
median_days_employed = df['days_employed'].median()
median_days_employed

2194.220566878695

In [213]:
dob_days_max = df['dob_years'].max() * 365
dob_days_max

27375

In [214]:
df.loc[df['days_employed'] >= 27375, 'days_employed'] = median_days_employed
df.sort_values(by='days_employed', ascending=False).head(20)

children  days_employed  dob_years            education  education_id  \
16335         1   18388.949901         61  secondary education             1   
4299          0   17615.563266         61  secondary education             1   
7329          0   16593.472817         60    bachelor's degree             0   
17838         0   16264.699501         59  secondary education             1   
16825         0   16119.687737         64  secondary education             1   
3974          0   15835.725775         64  secondary education             1   
1539          0   15785.678893         59    bachelor's degree             0   
4321          0   15773.061335         61  secondary education             1   
7731          0   15618.063786         64  secondary education             1   
15675         0   15410.040779         65    bachelor's degree             0   
19788         0   15267.541183         56  secondary education             1   
16868         0   15193.032201         57  secondary education             1   
5581          0   15079.216069         55  secondary education             1   
14908         0   14920.049805         56    bachelor's degree             0   
12969         0   14747.887934         68  secondary education             1   
8851          1   14677.902293         59  secondary education             1   
18116         0   14652.093786         59  secondary education             1   
12673         0   14583.152504         58  secondary education             1   
8982          0   14492.543987         60  secondary education             1   
18383         0   14447.661078         59  secondary education             1   

           family_status  family_status_id gender    income_type  debt  \
16335            married                 0      F       employee     0   
4299             married                 0      F       business     0   
7329             married                 0      F       employee     0   
17838            married                 0      F       employee     0   
16825            married                 0      F       employee     0   
3974   civil partnership                 1      F       business     0   
1539           unmarried                 4      F       employee     0   
4321   civil partnership                 1      F       employee     0   
7731             married                 0      F       business     0   
15675            married                 0      F       employee     0   
19788            married                 0      M       employee     0   
16868            married                 0      F  civil servant     0   
5581             married                 0      F  civil servant     0   
14908            married                 0      F       employee     0   
12969  civil partnership                 1      F       employee     0   
8851             married                 0      M       employee     0   
18116            married                 0      F  civil servant     0   
12673    widow / widower                 2      F  civil servant     0   
8982             married                 0      F       business     0   
18383            married                 0      F  civil servant     0   

       total_income                                   purpose  
16335     29788.629                  real estate transactions  
4299      19609.719                     purchase of the house  
7329      19951.655                       going to university  
17838      8198.235                              to buy a car  
16825     14644.430               buy residential real estate  
3974      15497.365                         to have a wedding  
1539      19130.216  transactions with commercial real estate  
4321      32938.974                          wedding ceremony  
7731      47444.057                      university education  
15675     30208.011               buy residential real estate  
19788     19676.219                              car purchase  
168

In [215]:
df['days_employed'].max() / 365

50.38068465909146

<div class="alert alert-danger">
<b>Reviewer's comment</b> <a class="tocSkip"></a>

- minta tolong untuk kolom days_employed dicek kembali ya. karena disitu ada user yg masa kerjanya 340.266 hari atau sekitar 932 tahun (?)

</div>

<div class="alert alert-block alert-info">
<b>Student answer.</b> <a class="tocSkip"></a>
    
Terima Kasih Untuk Mas Dipta untuk Reviewnya. Sudah diperbaiki, pertama kita lihat kembali maksimal 'dob_years' adalah 75 tahun atau 27375 hari. Jika ada nasabah yang masa kerjanya lebih dari itu maka kemungkinan ada salah penginputan atau salah format data. Maka untuk mengganti nilai tersebut kita gunakan median dari 'days_employed' yang bernilai 2194 hari. Setelah dilakukan pergantian nilai, untuk maksimal 'days_employed' adalah 50,3 tahun.
</div>

<div class="alert alert-block alert-info">Nasabah ada yang bernilai 0 Tahun.<a class="tocSkip"></a></div>

In [216]:
# Atasi masalah pada kolom `dob_years`, jika terdapat masalah

In [217]:
# Periksa `dob_years` untuk nilai yang mencurigakan dan hitung persentasenya
dob_years_count = df['dob_years'].count()
dob_years_zero = df[df['dob_years'] == 0]
dob_years_zero = dob_years_zero.shape[0]

dob_years_zero_percent = dob_years_zero / dob_years_count
dob_years_zero_percent
print('Jumlah percentage untuk nasabah yang berumur 0 Tahun adalah {:.2%}'.format(dob_years_zero_percent))



Jumlah percentage untuk nasabah yang berumur 0 Tahun adalah 0.47%


<div class="alert alert-block alert-info">Karena tujuan kita mencari tahu pengaruh status perkawinan seorang nasabah dan jumlah anak terhadap probabilitas gagal bayar dalam pinjaman, jumlah percentage untuk nasabah berumur 0 Tahun hanya  0.47% tidak terlalu berpengaruh. Ada kemungkinan juga hal tersebut terjadi karena salah penginputan. Jadi untuk saat ini kita biarkan saja, tidak menghapus baris - baris tersebut.   <a class="tocSkip"></a></div>

<div class="alert alert-success">
<b>Reviewer's comment</b> <a class="tocSkip"></a>

Bagus, semua berjalan lancar.

</div>

In [218]:
# Mari kita lihat nilai untuk kolom ini
df['family_status'].sort_values().unique()



array(['civil partnership', 'divorced', 'married', 'unmarried',
       'widow / widower'], dtype=object)

In [219]:
# Atasi nilai yang bermasalah di `family_status`, jika ada



In [220]:
# Periksa hasilnya - pastikan nilainya telah diperbaiki


<div class="alert alert-success">
<b>Reviewer's comment</b> <a class="tocSkip"></a>

Bagus, semua berjalan lancar.

</div>

In [221]:
# Mari kita liat nilai dalam kolom ini
df['gender'].sort_values().unique()

array(['F', 'M', 'XNA'], dtype=object)

In [222]:
# Atasi nilai-nilai yang bermasalah, jika ada
df = df.loc[df['gender'] != 'XNA']

In [223]:
# Periksa hasilnya - pastikan bahwa masalahnya telah diperbaiki
df['gender'].value_counts()

F    14236
M     7288
Name: gender, dtype: int64

<div class="alert alert-success">
<b>Reviewer's comment</b> <a class="tocSkip"></a>

Bagus, semua berjalan lancar.

</div>

In [224]:
# Mari kita lihat nilai dalam kolom ini
df['income_type'].sort_values().unique()

array(['business', 'civil servant', 'employee', 'entrepreneur',
       'paternity / maternity leave', 'retiree', 'student', 'unemployed'],
      dtype=object)

In [225]:
# Atasi nilai yang bermasalah, jika ada
df[ (df['income_type'] == 'student') | (df['income_type'] == 'unemployed')]

children  days_employed  dob_years            education  education_id  \
3133          1    2194.220567         31  secondary education             1   
9410          0     578.751554         22    bachelor's degree             0   
14798         0    2194.220567         45    bachelor's degree             0   

           family_status  family_status_id gender income_type  debt  \
3133             married                 0      M  unemployed     1   
9410           unmarried                 4      M     student     0   
14798  civil partnership                 1      F  unemployed     0   

       total_income                          purpose  
3133       9593.119  buying property for renting out  
9410      15712.260     construction of own property  
14798     32435.602               housing renovation

In [226]:
# Periksa hasilnya - pastikan bahwa masalahnya telah diperbaiki

<div class="alert alert-block alert-info">Walaupun beberapa nasabah ada yang 'income_type' nya student dan unemployed, namun beberapa dari mereka memiliki debt = 0 (Tidak memiliki hutang pembayaran pinjaman). Jadi untuk sat ini tidak ada yang perlu di perbaiki di kolom 'income_type'.<a class="tocSkip"></a></div>

<div class="alert alert-success">
<b>Reviewer's comment</b> <a class="tocSkip"></a>

Bagus, semua berjalan lancar.

</div>

In [227]:
# Periksa duplikat
df[df.duplicated()]


children  days_employed  dob_years            education  education_id  \
2849          0            NaN         41  secondary education             1   
3290          0            NaN         58  secondary education             1   
4182          1            NaN         34    bachelor's degree             0   
4851          0            NaN         60  secondary education             1   
5557          0            NaN         58  secondary education             1   
...         ...            ...        ...                  ...           ...   
20702         0            NaN         64  secondary education             1   
21032         0            NaN         60  secondary education             1   
21132         0            NaN         47  secondary education             1   
21281         1            NaN         30    bachelor's degree             0   
21415         0            NaN         54  secondary education             1   

           family_status  family_status_id gender income_type  debt  \
2849             married                 0      F    employee     0   
3290   civil partnership                 1      F     retiree     0   
4182   civil partnership                 1      F    employee     0   
4851   civil partnership                 1      F     retiree     0   
5557   civil partnership                 1      F     retiree     0   
...                  ...               ...    ...         ...   ...   
20702            married                 0      F     retiree     0   
21032            married                 0      F     retiree     0   
21132            married                 0      F    employee     0   
21281            married                 0      F    employee     0   
21415            married                 0      F     retiree     0   

       total_income                              purpose  
2849            NaN  purchase of the house for my family  
3290            NaN                    to have a wedding  
4182            NaN                     wedding ceremony  
4851            NaN                     wedding ceremony  
5557            NaN                    to have a wedding  
...             ...                                  ...  
20702           NaN              supplementary education  
21032           NaN                   to become educated  
21132           NaN                   housing renovation  
21281           NaN           buy commercial real estate  
21415           NaN                 housing transactions  

[71 rows x 12 columns]

In [228]:
# Atasi duplikat, jika ada
df_credit_analysis = df.drop_duplicates().reset_index(drop=True)

In [229]:
# Lakukan pemeriksaan terakhir untuk mengecek apakah kita memiliki duplikat
df_credit_analysis[df_credit_analysis.duplicated()]

Empty DataFrame
Columns: [children, days_employed, dob_years, education, education_id, family_status, family_status_id, gender, income_type, debt, total_income, purpose]
Index: []

In [230]:
# Periksa ukuran dataset yang sekarang Anda miliki setelah manipulasi pertama yang Anda lakukan
df.shape

(21524, 12)

In [231]:
df_credit_analysis.shape

(21453, 12)

In [232]:
df_shape1 = df.shape[0]
df_shape2 = df_credit_analysis.shape[0]
df_shape_percent =   1 - (int(df_shape2) * 1) / int(df_shape1)
df_shape_percent

0.003298643374837429

In [233]:
print('Jumlah percentage untuk perubahan ukuran dataset adalah {:.2%}'.format(df_shape_percent))

Jumlah percentage untuk perubahan ukuran dataset adalah 0.33%


<div class="alert alert-block alert-info">Jumlah percentage perubahan dengan duplikat yang dihapus masih sangat kecil, maka perlu dilakukan perbaikan pada nilai yang hilang.<a class="tocSkip"></a></div>

<div class="alert alert-success">
<b>Reviewer's comment</b> <a class="tocSkip"></a>

Bagus, semua berjalan lancar.

</div>

# Bekerja dengan nilai yang hilang

In [234]:
# Temukan dictionary
df_credit_analysis.describe()

children  days_employed     dob_years  education_id  \
count  21453.000000   19350.000000  21453.000000  21453.000000   
mean       0.544353    2324.744324     43.272130      0.817042   
std        1.381753    2090.015644     12.570427      0.548628   
min        0.000000      24.141633      0.000000      0.000000   
25%        0.000000     926.990457     33.000000      1.000000   
50%        0.000000    2194.218768     42.000000      1.000000   
75%        1.000000    2747.486276     53.000000      1.000000   
max       20.000000   18388.949901     75.000000      4.000000   

       family_status_id          debt   total_income  
count      21453.000000  21453.000000   19350.000000  
mean           0.973896      0.081154   26787.266688  
std            1.421601      0.273078   16475.822926  
min            0.000000      0.000000    3306.762000  
25%            0.000000      0.000000   16486.515250  
50%            0.000000      0.000000   23201.873500  
75%            1.000000      0.000000   32547.910750  
max            4.000000      1.000000  362496.645000

### Memperbaiki nilai yang hilang di `total_income`

<div class="alert alert-block alert-info">Seperti keimpulan sebelumnya, kolom 'days_employed' dan 'total_income' terdapat nilai yang hilang. Dengan jumlah nilai sekitar 10%, sebenarnya bisa saja baris-baris dengan nilai hilang tersebut bisa di hapus, namun bisa juga mengisi nilai hilang tersebut dengan nilai representatif.<a class="tocSkip"></a></div>


In [235]:
df_credit_analysis['dob_years'].sort_values().unique()

array([ 0, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34,
       35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51,
       52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67, 68,
       69, 70, 71, 72, 73, 74, 75])

In [236]:
# Mari kita tulis sebuah fungsi untuk menghitung kategori usia
def dob_years_group(age):
    if age < 0 or pd.isna(age):
        return 'NA'
    elif age < 10:
        return '0-9'
    elif age < 20:
        return '10-19' 
    elif age < 30:
        return '20-29'
    elif age < 40:
        return '30-39'
    elif age < 50:
        return '40-49'
    elif age < 60:
        return '50-59'
    elif age < 70:
        return '60-69'
    else:
        return '70+'

In [237]:
# Lakukan pengujian untuk melihat apakah fungsi Anda bekerja atau tidak
dob_years_group(69)


'60-69'

In [238]:
# Buatlah kolom baru berdasarkan fungsi
df_credit_analysis['dob_years_group'] = df_credit_analysis['dob_years'].apply(dob_years_group)


In [239]:
# Periksa bagaimana nilai di dalam kolom baru
df_credit_analysis.head(10)


children  days_employed  dob_years            education  education_id  \
0         1    8437.673028         42    bachelor's degree             0   
1         1    4024.803754         36  secondary education             1   
2         0    5623.422610         33  secondary education             1   
3         3    4124.747207         32  secondary education             1   
4         0    2194.220567         53  secondary education             1   
5         0     926.185831         27    bachelor's degree             0   
6         0    2879.202052         43    bachelor's degree             0   
7         0     152.779569         50  secondary education             1   
8         2    6929.865299         35    bachelor's degree             0   
9         0    2188.756445         41  secondary education             1   

       family_status  family_status_id gender income_type  debt  total_income  \
0            married                 0      F    employee     0     40620.102   
1            married                 0      F    employee     0     17932.802   
2            married                 0      M    employee     0     23341.752   
3            married                 0      M    employee     0     42820.568   
4  civil partnership                 1      F     retiree     0     25378.572   
5  civil partnership                 1      M    business     0     40922.170   
6            married                 0      F    business     0     38484.156   
7            married                 0      M    employee     0     21731.829   
8  civil partnership                 1      F    employee     0     15337.093   
9            married                 0      M    employee     0     23108.150   

                               purpose dob_years_group  
0                purchase of the house           40-49  
1                         car purchase           30-39  
2                purchase of the house           30-39  
3              supplementary education           30-39  
4                    to have a wedding           50-59  
5                purchase of the house           20-29  
6                 housing transactions           40-49  
7                            education           50-59  
8                     having a wedding           30-39  
9  purchase of the house for my family           40-49

In [240]:
#data_final.pivot_table(index='category_name', columns='source', values='visits', aggfunc='sum')
df_credit_analysis.pivot_table(index='dob_years_group', values='total_income', aggfunc=['median', 'mean'])

median          mean
                total_income  total_income
dob_years_group                           
0-9               24387.0700  25334.072890
10-19             14934.9010  16993.942462
20-29             22798.6650  25570.172966
30-39             24667.5280  28312.479963
40-49             24764.2290  28551.375635
50-59             22203.0745  25811.700327
60-69             19817.4400  23242.812818
70+               18751.3240  20125.658331

In [241]:
# Buat tabel tanpa nilai yang hilang dan tampilkan beberapa barisnya untuk memastikan semuanya berjalan dengan baik
#df_credit_analysis.info()
df_credit_analysis_notnull = df_credit_analysis[df_credit_analysis['total_income'].notnull()]
df_credit_analysis_notnull
df_credit_analysis_notnull.info()

print()
print("Informasi statistik deskriptif pada 'df_credit_analysis_notnull' dengan kolom 'total_income' :")
df_credit_analysis_notnull['total_income'].describe()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 19350 entries, 0 to 21452
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          19350 non-null  int64  
 1   days_employed     19350 non-null  float64
 2   dob_years         19350 non-null  int64  
 3   education         19350 non-null  object 
 4   education_id      19350 non-null  int64  
 5   family_status     19350 non-null  object 
 6   family_status_id  19350 non-null  int64  
 7   gender            19350 non-null  object 
 8   income_type       19350 non-null  object 
 9   debt              19350 non-null  int64  
 10  total_income      19350 non-null  float64
 11  purpose           19350 non-null  object 
 12  dob_years_group   19350 non-null  object 
dtypes: float64(2), int64(5), object(6)
memory usage: 2.1+ MB

Informasi statistik deskriptif pada 'df_credit_analysis_notnull' dengan kolom 'total_income' :


count     19350.000000
mean      26787.266688
std       16475.822926
min        3306.762000
25%       16486.515250
50%       23201.873500
75%       32547.910750
max      362496.645000
Name: total_income, dtype: float64

In [242]:
# Perhatikan nilai rata-rata untuk pendapatan berdasarkan faktor yang telah Anda identifikasi
df_credit_analysis_notnull['total_income'].describe()['mean']

26787.26668764858

In [243]:
# Perhatikan nilai median untuk pendapatan berdasarkan faktor yang telah Anda identifikasi
df_credit_analysis_notnull['total_income'].describe()['50%']

23201.8735

[Ulangi perbandingan tersebut untuk beberapa faktor. Pastikan Anda mempertimbangkan berbagai aspek dan menjelaskan proses berpikir Anda.]



In [244]:
# print(position.groupby('month')['level'].mean())
df_credit_analysis_notnull.groupby('gender')['total_income'].mean()

gender
F    24655.604757
M    30907.144369
Name: total_income, dtype: float64

In [245]:
df_credit_analysis_notnull.groupby('family_status')['total_income'].mean()

family_status
civil partnership    26692.840381
divorced             27189.354550
married              27041.784689
unmarried            26934.069805
widow / widower      22984.208556
Name: total_income, dtype: float64

In [246]:
df_credit_analysis_notnull.groupby('gender')['total_income'].median()

gender
F    21464.845
M    26834.295
Name: total_income, dtype: float64

In [247]:
df_credit_analysis_notnull.groupby('family_status')['total_income'].median()

family_status
civil partnership    23185.477
divorced             23515.096
married              23389.540
unmarried            23149.028
widow / widower      20514.190
Name: total_income, dtype: float64

<div class="alert alert-block alert-info">Berdasarkan metode describe()[50%] dapat diketahui mediannya adalah 23201, dan jika berdasarkan metode groupby('family_status') mediannya banyak muncul disekitaan angka 23K. Jadi kita gunakan angka median karena jika menggunakan mean rentangnya terlalu banyak.<a class="tocSkip"></a></div>


In [248]:
df_credit_analysis.groupby(['income_type'])['total_income'].median()

income_type
business                       27571.0825
civil servant                  24071.6695
employee                       22815.1035
entrepreneur                   79866.1030
paternity / maternity leave     8612.6610
retiree                        18962.3180
student                        15712.2600
unemployed                     21014.3605
Name: total_income, dtype: float64

In [249]:
#  Tulis fungsi yang akan kita gunakan untuk mengisi nilai yang hilang
business_median = df_credit_analysis['income_type'] == 'business'
df_credit_analysis.loc[business_median, 'total_income'] = df_credit_analysis.loc[business_median, 'total_income'].fillna(27571)

civil_median = df_credit_analysis['income_type'] == 'civil servant'
df_credit_analysis.loc[civil_median, 'total_income'] = df_credit_analysis.loc[civil_median, 'total_income'].fillna(24071)

employee_median = df_credit_analysis['income_type'] == 'employee'
df_credit_analysis.loc[employee_median, 'total_income'] = df_credit_analysis.loc[employee_median, 'total_income'].fillna(22815)

entrepreneur_median = df_credit_analysis['income_type'] == 'entrepreneur'
df_credit_analysis.loc[entrepreneur_median, 'total_income'] = df_credit_analysis.loc[entrepreneur_median, 'total_income'].fillna(79866)

leave_median = df_credit_analysis['income_type'] == 'paternity / maternity leave'
df_credit_analysis.loc[leave_median, 'total_income'] = df_credit_analysis.loc[leave_median, 'total_income'].fillna(8612)

retiree_median = df_credit_analysis['income_type'] == 'retiree'
df_credit_analysis.loc[retiree_median, 'total_income'] = df_credit_analysis.loc[retiree_median, 'total_income'].fillna(18962)

student_median = df_credit_analysis['income_type'] == 'student'
df_credit_analysis.loc[student_median, 'total_income'] = df_credit_analysis.loc[student_median, 'total_income'].fillna(15712)

unemployed_median = df_credit_analysis['income_type'] == 'unemployed'
df_credit_analysis.loc[unemployed_median, 'total_income'] = df_credit_analysis.loc[unemployed_median, 'total_income'].fillna(21014)


#df_credit_analysis.loc[unemployed_median, 'total_income'].isna().sum()
#df_credit_analysis[df_credit_analysis['total_income'] == 21014]


In [250]:
df_credit_analysis[df_credit_analysis['total_income'].isnull()]

Empty DataFrame
Columns: [children, days_employed, dob_years, education, education_id, family_status, family_status_id, gender, income_type, debt, total_income, purpose, dob_years_group]
Index: []

In [251]:
# Memeriksa bagaimana nilai di dalam kolom baru
#df_credit_analysis[df_credit_analysis['days_employed'].isna()]
df_credit_analysis[df_credit_analysis['total_income'].isna()]
#df_credit_analysis.info()
#df_credit_analysis.head(10)


Empty DataFrame
Columns: [children, days_employed, dob_years, education, education_id, family_status, family_status_id, gender, income_type, debt, total_income, purpose, dob_years_group]
Index: []

In [252]:
# Terapkan fungsi tersebut ke setiap baris

In [253]:
# Periksa apakah kita mendapatkan kesalahan
df_credit_analysis.describe()

children  days_employed     dob_years  education_id  \
count  21453.000000   19350.000000  21453.000000  21453.000000   
mean       0.544353    2324.744324     43.272130      0.817042   
std        1.381753    2090.015644     12.570427      0.548628   
min        0.000000      24.141633      0.000000      0.000000   
25%        0.000000     926.990457     33.000000      1.000000   
50%        0.000000    2194.218768     42.000000      1.000000   
75%        1.000000    2747.486276     53.000000      1.000000   
max       20.000000   18388.949901     75.000000      4.000000   

       family_status_id          debt   total_income  
count      21453.000000  21453.000000   21453.000000  
mean           0.973896      0.081154   26450.763269  
std            1.421601      0.273078   15710.271920  
min            0.000000      0.000000    3306.762000  
25%            0.000000      0.000000   17219.352000  
50%            0.000000      0.000000   22815.000000  
75%            1.000000      0.000000   31331.009000  
max            4.000000      1.000000  362496.645000

In [254]:
df_credit_analysis.groupby(['income_type'])['total_income'].mean()

income_type
business                       31911.522860
civil servant                  27018.028972
employee                       25530.671113
entrepreneur                   79866.051500
paternity / maternity leave     8612.661000
retiree                        21640.143712
student                        15712.260000
unemployed                     21014.360500
Name: total_income, dtype: float64

In [255]:
df_credit_analysis.groupby(['income_type'])['total_income'].median()

income_type
business                       27571.0000
civil servant                  24071.0000
employee                       22815.0000
entrepreneur                   79866.0515
paternity / maternity leave     8612.6610
retiree                        18962.0000
student                        15712.2600
unemployed                     21014.3605
Name: total_income, dtype: float64

<div class="alert alert-danger">
<b>Reviewer's comment</b> <a class="tocSkip"></a>

- disini diminta untuk menggunakan conditional fillna ya, jadi tidak straight forward seperti diatas.
- maksudnya adalah jika menggunakan income_type + median. maka proses fillna adalah sebagai berikut
    - ketika income_type == business, maka akan di fillna 25997.86600 dst dst.
- jadi ada if else nya
- jika bisa dibikin fungsi akan lebih baik

</div>

<div class="alert alert-block alert-info">
<b>Student answer.</b> <a class="tocSkip"></a>
    
Terima Kasih Untuk Mas Dipta untuk Reviewnya. Sudah diperbaiki. Untuk pengisian nilai kosong di kolom 'total_income', dicari dahulu berdasarkan kelompok 'income_type'. Karena jenis pekerjaan nasabah dengan pendapatan nasabah dengan pekerjaan tersebut saling berhubungan.
</div>

In [256]:
# Ganti nilai yang hilang jika terdapat kesalahan
df_credit_analysis[df_credit_analysis['days_employed'].isna()]

children  days_employed  dob_years            education  education_id  \
12            0            NaN         65  secondary education             1   
26            0            NaN         41  secondary education             1   
29            0            NaN         63  secondary education             1   
41            0            NaN         50  secondary education             1   
55            0            NaN         54  secondary education             1   
...         ...            ...        ...                  ...           ...   
21417         2            NaN         47  secondary education             1   
21423         1            NaN         50  secondary education             1   
21425         0            NaN         48    bachelor's degree             0   
21430         1            NaN         42  secondary education             1   
21438         2            NaN         28  secondary education             1   

           family_status  family_status_id gender    income_type  debt  \
12     civil partnership                 1      M        retiree     0   
26               married                 0      M  civil servant     0   
29             unmarried                 4      F        retiree     0   
41               married                 0      F  civil servant     0   
55     civil partnership                 1      F        retiree     1   
...                  ...               ...    ...            ...   ...   
21417            married                 0      M       business     0   
21423  civil partnership                 1      F       employee     0   
21425            married                 0      F       business     0   
21430            married                 0      F       employee     0   
21438            married                 0      F       employee     0   

       total_income                   purpose dob_years_group  
12          18962.0         to have a wedding           60-69  
26          24071.0                 education           40-49  
29          18962.0    building a real estate           60-69  
41          24071.0  second-hand car purchase           50-59  
55          18962.0         to have a wedding           50-59  
...             ...                       ...             ...  
21417       27571.0         purchase of a car           40-49  
21423       22815.0          wedding ceremony           50-59  
21425       27571.0       building a property           40-49  
21430       22815.0    building a real estate           40-49  
21438       22815.0              car purchase           20-29  

[2103 rows x 13 columns]

In [257]:
# Periksa jumlah entri di kolom
df_credit_analysis.isnull().sum()
#df_credit_analysis.shape


children               0
days_employed       2103
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income           0
purpose                0
dob_years_group        0
dtype: int64

In [258]:
df_credit_analysis['total_income'] = df_credit_analysis['total_income'].astype('int')

In [259]:
df_credit_analysis.info()
#df_credit_analysis.head(10)
#df_credit_analysis.describe()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21453 entries, 0 to 21452
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21453 non-null  int64  
 1   days_employed     19350 non-null  float64
 2   dob_years         21453 non-null  int64  
 3   education         21453 non-null  object 
 4   education_id      21453 non-null  int64  
 5   family_status     21453 non-null  object 
 6   family_status_id  21453 non-null  int64  
 7   gender            21453 non-null  object 
 8   income_type       21453 non-null  object 
 9   debt              21453 non-null  int64  
 10  total_income      21453 non-null  int64  
 11  purpose           21453 non-null  object 
 12  dob_years_group   21453 non-null  object 
dtypes: float64(1), int64(6), object(6)
memory usage: 2.1+ MB


###  Memperbaiki nilai di `days_employed`

In [260]:
# Distribusi median dari `days_employed` berdasarkan parameter yang Anda identifikasi
df_credit_analysis.groupby(['income_type'])['days_employed'].median()



income_type
business                       1546.333214
civil servant                  2689.368353
employee                       1574.202821
entrepreneur                    520.848083
paternity / maternity leave    3296.759962
retiree                        2194.220567
student                         578.751554
unemployed                     2194.220567
Name: days_employed, dtype: float64

In [261]:
# Distribusi rata-rata dari `days_employed` berdasarkan parameter yang Anda identifikasi
df_credit_analysis.groupby(['income_type'])['days_employed'].mean()

income_type
business                       2111.470404
civil servant                  3399.896902
employee                       2326.499216
entrepreneur                    520.848083
paternity / maternity leave    3296.759962
retiree                        2194.220567
student                         578.751554
unemployed                     2194.220567
Name: days_employed, dtype: float64

In [262]:
df_credit_analysis.describe()

children  days_employed     dob_years  education_id  \
count  21453.000000   19350.000000  21453.000000  21453.000000   
mean       0.544353    2324.744324     43.272130      0.817042   
std        1.381753    2090.015644     12.570427      0.548628   
min        0.000000      24.141633      0.000000      0.000000   
25%        0.000000     926.990457     33.000000      1.000000   
50%        0.000000    2194.218768     42.000000      1.000000   
75%        1.000000    2747.486276     53.000000      1.000000   
max       20.000000   18388.949901     75.000000      4.000000   

       family_status_id          debt   total_income  
count      21453.000000  21453.000000   21453.000000  
mean           0.973896      0.081154   26450.314921  
std            1.421601      0.273078   15710.264006  
min            0.000000      0.000000    3306.000000  
25%            0.000000      0.000000   17219.000000  
50%            0.000000      0.000000   22815.000000  
75%            1.000000      0.000000   31331.000000  
max            4.000000      1.000000  362496.000000

<div class="alert alert-block alert-info">Sebelum mengganti 'days_employed' yang bernilai diatas 27375 hari atau 75 tahun, berdasarkan metode describe() median-nya ~2194, dan mean-nya 66918. Dengan kondisi seperti ini berarti terdapat otulier yang signifikan maka lebih baik menggunakan nilai median.<a class="tocSkip"></a></div>

In [263]:
# Mari tulis fungsi yang menghitung rata-rata atau median (tergantung keputusan Anda) berdasarkan parameter yang Anda identifikasi
business_median_days = df_credit_analysis['income_type'] == 'business'
df_credit_analysis.loc[business_median_days, 'days_employed'] = df_credit_analysis.loc[business_median_days, 'days_employed'].fillna(1546)

civil_median_days = df_credit_analysis['income_type'] == 'civil servant'
df_credit_analysis.loc[civil_median_days, 'days_employed'] = df_credit_analysis.loc[civil_median_days, 'days_employed'].fillna(2689)

employee_median_days = df_credit_analysis['income_type'] == 'employee'
df_credit_analysis.loc[employee_median_days, 'days_employed'] = df_credit_analysis.loc[employee_median_days, 'days_employed'].fillna(1574)

entrepreneur_median_days = df_credit_analysis['income_type'] == 'entrepreneur'
df_credit_analysis.loc[entrepreneur_median_days, 'days_employed'] = df_credit_analysis.loc[entrepreneur_median_days, 'days_employed'].fillna(520)

leave_median_days = df_credit_analysis['income_type'] == 'paternity / maternity leave'
df_credit_analysis.loc[leave_median_days, 'days_employed'] = df_credit_analysis.loc[leave_median_days, 'days_employed'].fillna(3296)

retiree_median_days = df_credit_analysis['income_type'] == 'retiree'
df_credit_analysis.loc[retiree_median_days, 'days_employed'] = df_credit_analysis.loc[retiree_median_days, 'days_employed'].fillna(2194)

student_median_days = df_credit_analysis['income_type'] == 'student'
df_credit_analysis.loc[student_median_days, 'days_employed'] = df_credit_analysis.loc[student_median_days, 'days_employed'].fillna(1578)

unemployed_median_days = df_credit_analysis['income_type'] == 'unemployed'
df_credit_analysis.loc[unemployed_median_days, 'days_employed'] = df_credit_analysis.loc[unemployed_median_days, 'days_employed'].fillna(2194)


df_credit_analysis.loc[business_median_days, 'days_employed'].isna().sum()
df_credit_analysis[df_credit_analysis['days_employed'] == 1546]

children  days_employed  dob_years            education  education_id  \
65            0         1546.0         21  secondary education             1   
94            1         1546.0         34    bachelor's degree             0   
121           0         1546.0         29    bachelor's degree             0   
135           0         1546.0         27  secondary education             1   
174           0         1546.0         55    bachelor's degree             0   
...         ...            ...        ...                  ...           ...   
21319        20         1546.0         53  secondary education             1   
21320         0         1546.0         52  secondary education             1   
21336         1         1546.0         36  secondary education             1   
21417         2         1546.0         47  secondary education             1   
21425         0         1546.0         48    bachelor's degree             0   

           family_status  family_status_id gender income_type  debt  \
65             unmarried                 4      M    business     0   
94     civil partnership                 1      F    business     0   
121              married                 0      F    business     0   
135              married                 0      M    business     0   
174      widow / widower                 2      F    business     0   
...                  ...               ...    ...         ...   ...   
21319            married                 0      M    business     0   
21320            married                 0      F    business     0   
21336            married                 0      F    business     0   
21417            married                 0      M    business     0   
21425            married                 0      F    business     0   

       total_income                                   purpose dob_years_group  
65            27571  transactions with commercial real estate           20-29  
94            27571                          having a wedding           30-39  
121           27571                                       car           20-29  
135           27571                                   housing           20-29  
174           27571                              to own a car           50-59  
...             ...                                       ...             ...  
21319         27571               buy residential real estate           50-59  
21320         27571       purchase of the house for my family           50-59  
21336         27571                    building a real estate           30-39  
21417         27571                         purchase of a car           40-49  
21425         27571                       building a property           40-49  

[501 rows x 13 columns]

In [264]:
# Periksa apakah fungsi Anda dapat bekerja
df_credit_analysis.isnull().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dob_years_group     0
dtype: int64

In [265]:
# Terapkan fungsi ke income_type



In [266]:
# Periksa apakah fungsi Anda bekerja



In [267]:
# Ganti nilai yang hilang
#df_credit_analysis['days_employed'] = df_credit_analysis['days_employed'].fillna(median_days_employed)


<div class="alert alert-danger">
<b>Reviewer's comment</b> <a class="tocSkip"></a>

- disini diminta untuk menggunakan conditional fillna ya, jadi tidak straight forward seperti diatas.
- seperti revisi diatasnya
- jadi ada if else nya
- - jika bisa dibikin fungsi akan lebih baik

</div>

<div class="alert alert-block alert-info">
<b>Student answer.</b> <a class="tocSkip"></a>
    
Terima Kasih Untuk Mas Dipta untuk Reviewnya. Sudah diperbaiki.
</div>

In [268]:
# Periksa entri di semua kolom - pastikan kita memperbaiki semua nilai yang hilang
df_credit_analysis.isnull().sum()
df_credit_analysis.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21453 entries, 0 to 21452
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21453 non-null  int64  
 1   days_employed     21453 non-null  float64
 2   dob_years         21453 non-null  int64  
 3   education         21453 non-null  object 
 4   education_id      21453 non-null  int64  
 5   family_status     21453 non-null  object 
 6   family_status_id  21453 non-null  int64  
 7   gender            21453 non-null  object 
 8   income_type       21453 non-null  object 
 9   debt              21453 non-null  int64  
 10  total_income      21453 non-null  int64  
 11  purpose           21453 non-null  object 
 12  dob_years_group   21453 non-null  object 
dtypes: float64(1), int64(6), object(6)
memory usage: 2.1+ MB


In [269]:
df_credit_analysis['days_employed'] = df_credit_analysis['days_employed'].astype('int')

In [270]:
df_credit_analysis.info()
df_credit_analysis.head(10)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21453 entries, 0 to 21452
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   children          21453 non-null  int64 
 1   days_employed     21453 non-null  int64 
 2   dob_years         21453 non-null  int64 
 3   education         21453 non-null  object
 4   education_id      21453 non-null  int64 
 5   family_status     21453 non-null  object
 6   family_status_id  21453 non-null  int64 
 7   gender            21453 non-null  object
 8   income_type       21453 non-null  object
 9   debt              21453 non-null  int64 
 10  total_income      21453 non-null  int64 
 11  purpose           21453 non-null  object
 12  dob_years_group   21453 non-null  object
dtypes: int64(7), object(6)
memory usage: 2.1+ MB


children  days_employed  dob_years            education  education_id  \
0         1           8437         42    bachelor's degree             0   
1         1           4024         36  secondary education             1   
2         0           5623         33  secondary education             1   
3         3           4124         32  secondary education             1   
4         0           2194         53  secondary education             1   
5         0            926         27    bachelor's degree             0   
6         0           2879         43    bachelor's degree             0   
7         0            152         50  secondary education             1   
8         2           6929         35    bachelor's degree             0   
9         0           2188         41  secondary education             1   

       family_status  family_status_id gender income_type  debt  total_income  \
0            married                 0      F    employee     0         40620   
1            married                 0      F    employee     0         17932   
2            married                 0      M    employee     0         23341   
3            married                 0      M    employee     0         42820   
4  civil partnership                 1      F     retiree     0         25378   
5  civil partnership                 1      M    business     0         40922   
6            married                 0      F    business     0         38484   
7            married                 0      M    employee     0         21731   
8  civil partnership                 1      F    employee     0         15337   
9            married                 0      M    employee     0         23108   

                               purpose dob_years_group  
0                purchase of the house           40-49  
1                         car purchase           30-39  
2                purchase of the house           30-39  
3              supplementary education           30-39  
4                    to have a wedding           50-59  
5                purchase of the house           20-29  
6                 housing transactions           40-49  
7                            education           50-59  
8                     having a wedding           30-39  
9  purchase of the house for my family           40-49

In [271]:
df_credit_analysis.describe()

children  days_employed     dob_years  education_id  \
count  21453.000000   21453.000000  21453.000000  21453.000000   
mean       0.544353    2268.733557     43.272130      0.817042   
std        1.381753    1995.103503     12.570427      0.548628   
min        0.000000      24.000000      0.000000      0.000000   
25%        0.000000    1023.000000     33.000000      1.000000   
50%        0.000000    1996.000000     42.000000      1.000000   
75%        1.000000    2575.000000     53.000000      1.000000   
max       20.000000   18388.000000     75.000000      4.000000   

       family_status_id          debt   total_income  
count      21453.000000  21453.000000   21453.000000  
mean           0.973896      0.081154   26450.314921  
std            1.421601      0.273078   15710.264006  
min            0.000000      0.000000    3306.000000  
25%            0.000000      0.000000   17219.000000  
50%            0.000000      0.000000   22815.000000  
75%            1.000000      0.000000   31331.000000  
max            4.000000      1.000000  362496.000000

## Pengkategorian Data

In [272]:
# Tampilkan nilai data yang Anda pilih untuk pengkategorian
df_credit_analysis[['total_income', 'purpose']]


total_income                  purpose
0             40620    purchase of the house
1             17932             car purchase
2             23341    purchase of the house
3             42820  supplementary education
4             25378        to have a wedding
...             ...                      ...
21448         35966     housing transactions
21449         24959        purchase of a car
21450         14347                 property
21451         39054        buying my own car
21452         13127             to buy a car

[21453 rows x 2 columns]

In [273]:
# Periksa nilai unik
df_credit_analysis['purpose'].sort_values().unique()

array(['building a property', 'building a real estate',
       'buy commercial real estate', 'buy real estate',
       'buy residential real estate', 'buying a second-hand car',
       'buying my own car', 'buying property for renting out', 'car',
       'car purchase', 'cars', 'construction of own property',
       'education', 'getting an education', 'getting higher education',
       'going to university', 'having a wedding', 'housing',
       'housing renovation', 'housing transactions', 'profile education',
       'property', 'purchase of a car', 'purchase of my own house',
       'purchase of the house', 'purchase of the house for my family',
       'real estate transactions', 'second-hand car purchase',
       'supplementary education', 'to become educated', 'to buy a car',
       'to get a supplementary education', 'to have a wedding',
       'to own a car', 'transactions with commercial real estate',
       'transactions with my real estate', 'university education',
       'we

In [274]:
# Mari kita tulis sebuah fungsi untuk mengategorikan data berdasarkan topik umum
def purpose_group(purpose_list):
    if purpose_list in ['supplementary education', 'university education', 'education', 'to get a supplementary education', 'getting an education', 
     'getting higher education', 'profile education', 'to become educated', 'going to university']:
        return 'education'
    elif purpose_list in ['wedding', 'to have a wedding', 'having a wedding', 'wedding ceremony']:
        return 'wedding'
    elif purpose_list in ['real estate transactions', 'buy commercial real estate', 'building a real estate', 'transactions with commercial real estate', 
     'transactions with my real estate', 'buy real estate', 'buy residential real estate']:
        return 'real_estate' 
    elif purpose_list in ['buying property for renting out', 'construction of own property', 'property', 'building a property']:
        return 'property'
    elif purpose_list in  ['purchase of the house', 'housing', 'housing transactions', 'purchase of my own house', 
     'purchase of the house for my family', 'housing renovation']:
        return 'house'
    elif purpose_list in ['car', 'buying my own car', 'to buy a car', 'buying a second-hand car', 'second-hand car purchase', 'to own a car', 
     'cars', 'car purchase', 'purchase of a car']:
        return 'car'



In [275]:
# Buat kolom yang memuat kategori dan hitung nilainya
df_credit_analysis['purpose_group'] = df_credit_analysis['purpose'].apply(purpose_group)



In [276]:
df_credit_analysis['purpose_group'].sort_values().unique()

array(['car', 'education', 'house', 'property', 'real_estate', 'wedding'],
      dtype=object)

In [277]:
df_credit_analysis[df_credit_analysis['purpose_group'].isna()]
df_credit_analysis.info()
df_credit_analysis.head(10)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21453 entries, 0 to 21452
Data columns (total 14 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   children          21453 non-null  int64 
 1   days_employed     21453 non-null  int64 
 2   dob_years         21453 non-null  int64 
 3   education         21453 non-null  object
 4   education_id      21453 non-null  int64 
 5   family_status     21453 non-null  object
 6   family_status_id  21453 non-null  int64 
 7   gender            21453 non-null  object
 8   income_type       21453 non-null  object
 9   debt              21453 non-null  int64 
 10  total_income      21453 non-null  int64 
 11  purpose           21453 non-null  object
 12  dob_years_group   21453 non-null  object
 13  purpose_group     21453 non-null  object
dtypes: int64(7), object(7)
memory usage: 2.3+ MB


children  days_employed  dob_years            education  education_id  \
0         1           8437         42    bachelor's degree             0   
1         1           4024         36  secondary education             1   
2         0           5623         33  secondary education             1   
3         3           4124         32  secondary education             1   
4         0           2194         53  secondary education             1   
5         0            926         27    bachelor's degree             0   
6         0           2879         43    bachelor's degree             0   
7         0            152         50  secondary education             1   
8         2           6929         35    bachelor's degree             0   
9         0           2188         41  secondary education             1   

       family_status  family_status_id gender income_type  debt  total_income  \
0            married                 0      F    employee     0         40620   
1            married                 0      F    employee     0         17932   
2            married                 0      M    employee     0         23341   
3            married                 0      M    employee     0         42820   
4  civil partnership                 1      F     retiree     0         25378   
5  civil partnership                 1      M    business     0         40922   
6            married                 0      F    business     0         38484   
7            married                 0      M    employee     0         21731   
8  civil partnership                 1      F    employee     0         15337   
9            married                 0      M    employee     0         23108   

                               purpose dob_years_group purpose_group  
0                purchase of the house           40-49         house  
1                         car purchase           30-39           car  
2                purchase of the house           30-39         house  
3              supplementary education           30-39     education  
4                    to have a wedding           50-59       wedding  
5                purchase of the house           20-29         house  
6                 housing transactions           40-49         house  
7                            education           50-59     education  
8                     having a wedding           30-39       wedding  
9  purchase of the house for my family           40-49         house

<div class="alert alert-success">
<b>Reviewer's comment</b> <a class="tocSkip"></a>

Bagus, semua berjalan lancar.

</div>

In [278]:
# Lihat semua data numerik di kolom yang Anda pilih untuk pengkategorian
df_credit_analysis['total_income'].sort_values(ascending=False)

12389    362496
19547    352136
9159     276204
20741    274402
17136    273809
          ...  
14246      3503
1598       3471
16137      3418
12982      3392
14554      3306
Name: total_income, Length: 21453, dtype: int64

In [279]:
# Dapatkan kesimpulan statistik untuk kolomnya
df_credit_analysis['total_income'].describe()



count     21453.000000
mean      26450.314921
std       15710.264006
min        3306.000000
25%       17219.000000
50%       22815.000000
75%       31331.000000
max      362496.000000
Name: total_income, dtype: float64

<div class="alert alert-block alert-info">Berdasarkan statitiska deskripsi diatas untuk kolom 'total_income' nilai minimumnya adalah 3300 dan maksimumnya adalah 362496. Jadi kita akan membuat level 'total_income' dengan rentang 80000, dengan level terendah very_low. Hal ini untuk memudahkan dalam menganalisis hubungan antara total_income dengan pelunasan tepat waktu.<a class="tocSkip"></a></div>

In [280]:
# Buat fungsi yang melakukan pengkategorian menjadi kelompok numerik yang berbeda berdasarkan rentang
def total_income_level(range_income):
    if range_income < 80000:
        return 'very_low'
    elif range_income <= 160000:
        return 'low'
    elif range_income <= 240000:
        return 'medium'
    elif range_income <= 320000:
        return 'high'
    elif range_income > 320000:
        return 'very_high'
    
    
print(total_income_level(230000))


medium


In [281]:
# Buat kolom yang memuat kategori
df_credit_analysis['total_income_level'] = df_credit_analysis['total_income'].apply(total_income_level)

In [282]:
df_credit_analysis.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21453 entries, 0 to 21452
Data columns (total 15 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   children            21453 non-null  int64 
 1   days_employed       21453 non-null  int64 
 2   dob_years           21453 non-null  int64 
 3   education           21453 non-null  object
 4   education_id        21453 non-null  int64 
 5   family_status       21453 non-null  object
 6   family_status_id    21453 non-null  int64 
 7   gender              21453 non-null  object
 8   income_type         21453 non-null  object
 9   debt                21453 non-null  int64 
 10  total_income        21453 non-null  int64 
 11  purpose             21453 non-null  object
 12  dob_years_group     21453 non-null  object
 13  purpose_group       21453 non-null  object
 14  total_income_level  21453 non-null  object
dtypes: int64(7), object(8)
memory usage: 2.5+ MB


In [283]:
# Hitung setiap nilai kategori untuk melihat pendistribusiannya
df_credit_analysis.groupby('total_income_level')['total_income_level'].count()

total_income_level
high             5
low            197
medium          18
very_high        2
very_low     21231
Name: total_income_level, dtype: int64

In [284]:
df_credit_analysis.head(10)

children  days_employed  dob_years            education  education_id  \
0         1           8437         42    bachelor's degree             0   
1         1           4024         36  secondary education             1   
2         0           5623         33  secondary education             1   
3         3           4124         32  secondary education             1   
4         0           2194         53  secondary education             1   
5         0            926         27    bachelor's degree             0   
6         0           2879         43    bachelor's degree             0   
7         0            152         50  secondary education             1   
8         2           6929         35    bachelor's degree             0   
9         0           2188         41  secondary education             1   

       family_status  family_status_id gender income_type  debt  total_income  \
0            married                 0      F    employee     0         40620   
1            married                 0      F    employee     0         17932   
2            married                 0      M    employee     0         23341   
3            married                 0      M    employee     0         42820   
4  civil partnership                 1      F     retiree     0         25378   
5  civil partnership                 1      M    business     0         40922   
6            married                 0      F    business     0         38484   
7            married                 0      M    employee     0         21731   
8  civil partnership                 1      F    employee     0         15337   
9            married                 0      M    employee     0         23108   

                               purpose dob_years_group purpose_group  \
0                purchase of the house           40-49         house   
1                         car purchase           30-39           car   
2                purchase of the house           30-39         house   
3              supplementary education           30-39     education   
4                    to have a wedding           50-59       wedding   
5                purchase of the house           20-29         house   
6                 housing transactions           40-49         house   
7                            education           50-59     education   
8                     having a wedding           30-39       wedding   
9  purchase of the house for my family           40-49         house   

  total_income_level  
0           very_low  
1           very_low  
2           very_low  
3           very_low  
4           very_low  
5           very_low  
6           very_low  
7           very_low  
8           very_low  
9           very_low

<div class="alert alert-danger">
<b>Reviewer's comment</b> <a class="tocSkip"></a>

- kategori untuk income_type terlalu banyak yaitu 11 kategori. jadi sebisa mungkin max 5 kategori aja. bisa pakai income low, medium, high dsb dsb

</div>

<div class="alert alert-block alert-info">
<b>Student answer.</b> <a class="tocSkip"></a>
    
Terima Kasih Untuk Mas Dipta untuk Reviewnya. Sudah diperbaiki.
</div>

## Memeriksa hipotesis


**Apakah terdapat korelasi antara memiliki anak dengan melakukan pelunasan tepat waktu?**

In [285]:
# Periksa data anak dan data pelunasan tepat waktu
debt_by_children = df_credit_analysis.pivot_table(index='children', values='debt', aggfunc=['mean', 'sum', 'count', len])
debt_by_children
# Hitung gagal bayar berdasarkan jumlah anak

mean   sum  count    len
              debt  debt   debt   debt
children                              
0         0.075444  1063  14090  14090
1         0.091658   445   4855   4855
2         0.094542   194   2052   2052
3         0.081818    27    330    330
4         0.097561     4     41     41
5         0.000000     0      9      9
20        0.105263     8     76     76

In [286]:
zero_child = int(debt_by_children.loc[0,'mean'] * 100)
print('Rata-rata percentage debt untuk nasabah dengan jumlah anak 0 adalah ~',zero_child,'%')

Rata-rata percentage debt untuk nasabah dengan jumlah anak 0 adalah ~ 7 %


In [287]:
two_child = int(debt_by_children.loc[2,'mean'] * 100)
print('Rata-rata percentage debt untuk nasabah dengan jumlah anak 2 adalah ~',two_child,'%')

Rata-rata percentage debt untuk nasabah dengan jumlah anak 2 adalah ~ 9 %


In [288]:
three_child = int(debt_by_children.loc[3,'mean'] * 100)
print('Rata-rata percentage debt untuk nasabah dengan jumlah anak 3 adalah ~',three_child,'%')

Rata-rata percentage debt untuk nasabah dengan jumlah anak 3 adalah ~ 8 %


In [289]:
twenty_child = int(debt_by_children.loc[20,'mean'] * 100)
print('Rata-rata percentage debt untuk nasabah dengan jumlah anak 20 adalah ~',twenty_child,'%')

Rata-rata percentage debt untuk nasabah dengan jumlah anak 20 adalah ~ 10 %


**Kesimpulan**
<div class="alert alert-block alert-info">Berdasarkan data pivot 'debt_by_children', rata-rata debt dengan nasabah yang tidak mempunyai anak adalah ~0.07, dari 1 - 2 anak meningkat menjadi ~0.09. Kemudian untuk 3 anak menurun menjadi ~0.08, terakhir untuk 20 anak meningkat lagi menjadi ~0.1. Kesimpulannya korelasi antara memiliki anak dengan melakukan pelunasan tepat waktu tidak kuat. <a class="tocSkip"></a></div>


<div class="alert alert-success">
<b>Reviewer's comment</b> <a class="tocSkip"></a>

Bagus, semua berjalan lancar.

</div>

**Apakah terdapat korelasi antara status keluarga dengan pelunasan tepat waktu?**

In [290]:
# Periksa data status keluarga dan pelunasan tepat waktu

debt_by_family_status = df_credit_analysis.pivot_table(index='family_status', values='debt', aggfunc=['mean', 'sum', 'count', len])
debt_by_family_status

# Hitung gagal bayar berdasarkan status keluarga



mean  sum  count    len
                       debt debt   debt   debt
family_status                                 
civil partnership  0.093494  388   4150   4150
divorced           0.071130   85   1195   1195
married            0.075452  931  12339  12339
unmarried          0.097509  274   2810   2810
widow / widower    0.065693   63    959    959

**Kesimpulan**
<div class="alert alert-block alert-info">Berdasarkan pivot 'debt_by_family_status', debt pada unmarried lebih tinggi daripada married, dan menempati posisi tertinggi daripada status lainnya. Untuk status widow / widower paling rendah debtnya. <a class="tocSkip"></a></div>


**Apakah terdapat korelasi antara tingkat pendapatan dengan membayar kembali tepat waktu?**

In [291]:
# Periksa data tingkat pendapatan dan pelunasan tepat waktu

debt_by_total_income = df_credit_analysis.pivot_table(index='total_income_level', values='debt', aggfunc=['mean', 'sum', 'count', len])
debt_by_total_income

# Hitung gagal bayar berdasarkan tingkat pendapatan



mean   sum  count    len
                        debt  debt   debt   debt
total_income_level                              
high                0.000000     0      5      5
low                 0.060914    12    197    197
medium              0.055556     1     18     18
very_high           0.500000     1      2      2
very_low            0.081343  1727  21231  21231

**Kesimpulan**
<div class="alert alert-block alert-info">Berdasarkan pivot 'debt_by_total_income', korelasi antara tingkat pendapatan dengan pelunasan tepat waktu tidak kuat karena bisa dilihat rata-rata debt dari total_income_level very_low ke very_high naik turun, tidak ada polanya. Namun juga bisa diambil kesimpulan level very_high income yaitu range total_income diatas 320000 dari dua nasabah yang berhutang, salah satunya telat bayar hutang.. <a class="tocSkip"></a></div>

<div class="alert alert-success">
<b>Reviewer's comment</b> <a class="tocSkip"></a>

Bagus, semua berjalan lancar.

</div>

**Bagaimana tujuan kredit memengaruhi tingkat gagal bayar?**

In [292]:
# Periksa persentase tingkat gagal bayar untuk setiap tujuan kredit dan lakukan penganalisisan
debt_by_purpose = df_credit_analysis.pivot_table(index='purpose_group', values='debt', aggfunc=['mean', 'sum', 'count', len])
debt_by_purpose


mean  sum count   len
                   debt debt  debt  debt
purpose_group                           
car            0.093590  403  4306  4306
education      0.092200  370  4013  4013
house          0.067209  256  3809  3809
property       0.074862  190  2538  2538
real_estate    0.075286  336  4463  4463
wedding        0.080034  186  2324  2324

**Kesimpulan**
<div class="alert alert-block alert-info">Berdasarkan pivot 'debt_by_purpose' dapat disimpulkan bahwa nasabah yang meminjam untuk tujuan pembelian mobil dan pendidikan mempunyai debt yang lebih tinggi daripada tujuan lainnya. Dan untuk rumah debtnya paling rendah.<a class="tocSkip"></a></div>


<div class="alert alert-success">
<b>Reviewer's comment</b> <a class="tocSkip"></a>

Bagus, semua berjalan lancar.

</div>

# Kesimpulan umum 
<div class="alert alert-block alert-info">Di projek 'Menganalisis Risiko Gagal Bayar Peminjam' ini, kita mendapatkan data yang belum rapih.<br>
Kita menganalisa masalah informasi awal di data tersebut, contohnya.
- Kolom 'days_employed' masih banyak yang bernilai negatif. Ada 1 baris yang bernilai 'NaN'. Dan semua baris masih bernilai float.
- Kolom 'education' mempunya nilai duplikat, contohnya mengandung arti yang sama namum beda penulisan (Huruf Besar &    Huruf Kecil).
- Kolom 'total_income' ada 1 baris yang bernilai 'NaN' dan sejajar dengan 'NaN' di 'days_employed'. Semua baris masih bernilai float.</br>

Hal ini dapat kita atasi dengan beberapa metode dan fungsi.
<br><br>Untuk masalah nilai hilang dan duplikat yang ada di data tersebut dan langkah perbaikannya.</br></br>

- Nilai yang hilang di kolom kuantitatif berhasil diatasi dengan mengisi nilai representatif yaitu median dari kolom tersebuut.
-Nilai yang duplikat berhasil diatasi dengan mencari keunikan nilai dan mengkategorikannya.<a class="tocSkip"></a></div>

<div class="alert alert-block alert-info"><br>Dari 4 pivot tabel yang dibuat pada bagian 'Memeriksa Hipotesis' dapat disimpulkan :</br>
    
- Korelasi antara jumlah anak dengan waktu pelunasan hutang tidak kuat, namun nasabah dengan jumlah anak 20 terkendala membayar hutang tepat waktu daripada nasabah jumlah anak dibawah 20.
- Nasabah yang belum menikah lebih telat membayar hutang daripada yang menikah. Untuk nasabah janda atau duda paling rajin membayar utang tepat waktu.
- Korelasi antara total pendapat dengan waktu pelunasan hutang tidak kuat, namun nasabah dengan total pendapatan 120,000 sd / 150,000 paling telat membayar hutang daripada nasabah dengan total pendapatan dibawah 120,000 atau diatas 150,000.
- Nasabah yang melakukan pinjaman untuk pembelian mobil menempati nomor 1 untuk telat bayar hutang, dan di posisi 2 adalah nasabah dengan tujuan biaya pendidikan.<a class="tocSkip"></a></div>


<div class="alert alert-block alert-warning">

# General comments #1
<b>Reviewer's comment</b> <a class="tocSkip"></a> 

Good Job Mufti! Sebagian besar kamu telah menyelesaikan project ini dengan baik namun masih ada beberapa hal yang bisa diimprove yang saya tandai dengan comment berwarna kuning dan merah. Jika kamu telah merevisinya, projectmu akan di approve dan kamu bisa lanjut ke tahapan selanjutnya!.
    
Keep going Mufti, you are doing very well!
    

</div>

<div class="alert alert-block alert-success">

# General comments #2
<b>Reviewer's comment</b> <a class="tocSkip"></a> 

Good Job Mufti! Sudah melakukan revisi dengan baik dan bisa ke tahapan selanjutnya yah!
    

</div>